{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyNx/5f2OyB9v659bbI+AZLF"
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {
        "id": "krIK_AtSBq55"
      },
      "outputs": [],
      "source": [
        "from typing_extensions import Counter\n",
        "import collections\n",
        "import random\n",
        "\n",
        "nucleodites =['A','T','C','G']\n",
        "reverse_nucleodites = {'A':'T','T':'A','G':'C','C':'G'}\n",
        "\n",
        "def validate_seg_dna(dna):\n",
        "    verification = dna.upper()\n",
        "    for nuc in verification:\n",
        "        if nuc not in nucleodites:\n",
        "            return False\n",
        "    return verification\n",
        "\n",
        "def count_nucleodites(dna):\n",
        "    nucleodites_count = {'A':0,'T':0,'C':0,'G':0}\n",
        "    for nuc in dna:\n",
        "        nucleodites_count[nuc] += 1\n",
        "    return nucleodites_count\n",
        "\n",
        "def transcription(dna):\n",
        "    return dna.replace('T','U')\n",
        "\n",
        "def reverse_complement(dna):\n",
        "    #return ''.join([reverse_nucleodites[nuc] for nuc in dna])[::-1]\n",
        "    mapping=str(dna).maketrans('ATCG','TAGC')\n",
        "    return dna.translate(mapping)[::-1]\n",
        "\n",
        "\n",
        "def dna_complement(dna):\n",
        "    return ''.join([reverse_nucleodites[nuc] for nuc in dna])\n",
        "\n",
        "def dna_colors(dna):\n",
        "    dnacolors = {\n",
        "        'A': '\\033[92m',\n",
        "        'C': '\\033[94m',\n",
        "        'G': '\\033[93m',\n",
        "        'T': '\\033[91m',\n",
        "        'U': '\\033[91m',\n",
        "        'reset': '\\033[0;0m'\n",
        "    }\n",
        "\n",
        "    variavel = \"\"\n",
        "\n",
        "    for nuc in dna:\n",
        "        if nuc in dnacolors:\n",
        "            variavel += dnacolors[nuc] + nuc\n",
        "        else:\n",
        "            variavel += dnacolors['reset'] + nuc\n",
        "\n",
        "    return variavel + '\\033[0;0m'\n",
        "\n",
        "def gc_content(dna):\n",
        "  if len(dna) == 0:\n",
        "    return 0\n",
        "  return round((dna.count('C') + dna.count('G')) / len(dna) * 100,6)\n",
        "\n",
        "def gc_content_subsets(dna,k=20):\n",
        "  props=[]\n",
        "  for i in range(0,len(dna)-k+1,k):\n",
        "    subsets=dna[i:i+k]\n",
        "    props.append(gc_content(subsets))\n",
        "  return props\n",
        "\n",
        "\n",
        "def readTextFile(filePath):\n",
        "    with open(filePath, 'r') as f:\n",
        "        return \"\".join([l.strip() for l in f.readlines()])\n",
        "\n",
        "\n",
        "def writeTextFile(filePath, dna, mode='w'):\n",
        "    with open(filePath, mode) as f:\n",
        "        f.write(dna + '\\n')\n",
        "\n",
        "\n",
        "def read_FASTA(filePath):\n",
        "    with open(filePath, 'r') as f:\n",
        "        FASTAFile = [l.strip() for l in f.readlines()]\n",
        "\n",
        "    FASTADict = {}\n",
        "    FASTALabel = \"\"\n",
        "\n",
        "    for line in FASTAFile:\n",
        "        if '>' in line:\n",
        "            FASTALabel = line\n",
        "            FASTADict[FASTALabel] = \"\"\n",
        "        else:\n",
        "            FASTADict[FASTALabel] += line\n",
        "\n",
        "    return FASTADict\n",
        "\n",
        "# List the amoni acids\n",
        "def translate_dna(dna):\n",
        "    return [DNA_Codons[dna[pos:pos + 3]] for pos in range(0, len(dna)-2, 3)]\n",
        "\n",
        "#Frequency of a codon and returns the amino acid\n",
        "def codon_freq(dna,animoacid):\n",
        "  lista=[]\n",
        "  for i in range(0,len(dna)-2,3):\n",
        "    if DNA_Codons[dna[i:i+3]]==animoacid:\n",
        "      lista.append(dna[i:i+3])\n",
        "  freqdic=dict(Counter(lista))\n",
        "  total_freq=sum(freqdic.values())\n",
        "  for dna in freqdic:\n",
        "    freqdic[dna]=round(freqdic[dna]/total_freq,2)\n",
        "  return freqdic\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "NUCLEOTIDE_BASE = {\n",
        "    \"DNA\": [\"A\", \"T\", \"C\", \"G\"],\n",
        "    \"RNA\": [\"A\", \"U\", \"C\", \"G\"]\n",
        "}\n",
        "\n",
        "DNA_Codons = {\n",
        "    # 'M' - START, '_' - STOP\n",
        "    \"GCT\": \"A\", \"GCC\": \"A\", \"GCA\": \"A\", \"GCG\": \"A\",\n",
        "    \"TGT\": \"C\", \"TGC\": \"C\",\n",
        "    \"GAT\": \"D\", \"GAC\": \"D\",\n",
        "    \"GAA\": \"E\", \"GAG\": \"E\",\n",
        "    \"TTT\": \"F\", \"TTC\": \"F\",\n",
        "    \"GGT\": \"G\", \"GGC\": \"G\", \"GGA\": \"G\", \"GGG\": \"G\",\n",
        "    \"CAT\": \"H\", \"CAC\": \"H\",\n",
        "    \"ATA\": \"I\", \"ATT\": \"I\", \"ATC\": \"I\",\n",
        "    \"AAA\": \"K\", \"AAG\": \"K\",\n",
        "    \"TTA\": \"L\", \"TTG\": \"L\", \"CTT\": \"L\", \"CTC\": \"L\", \"CTA\": \"L\", \"CTG\": \"L\",\n",
        "    \"ATG\": \"M\",\n",
        "    \"AAT\": \"N\", \"AAC\": \"N\",\n",
        "    \"CCT\": \"P\", \"CCC\": \"P\", \"CCA\": \"P\", \"CCG\": \"P\",\n",
        "    \"CAA\": \"Q\", \"CAG\": \"Q\",\n",
        "    \"CGT\": \"R\", \"CGC\": \"R\", \"CGA\": \"R\", \"CGG\": \"R\", \"AGA\": \"R\", \"AGG\": \"R\",\n",
        "    \"TCT\": \"S\", \"TCC\": \"S\", \"TCA\": \"S\", \"TCG\": \"S\", \"AGT\": \"S\", \"AGC\": \"S\",\n",
        "    \"ACT\": \"T\", \"ACC\": \"T\", \"ACA\": \"T\", \"ACG\": \"T\",\n",
        "    \"GTT\": \"V\", \"GTC\": \"V\", \"GTA\": \"V\", \"GTG\": \"V\",\n",
        "    \"TGG\": \"W\",\n",
        "    \"TAT\": \"Y\", \"TAC\": \"Y\",\n",
        "    \"TAA\": \"_\", \"TAG\": \"_\", \"TGA\": \"_\"\n",
        "}\n",
        "\n",
        "RNA_Codons = {\n",
        "    # 'M' - START, '_' - STOP\n",
        "    \"GCU\": \"A\", \"GCC\": \"A\", \"GCA\": \"A\", \"GCG\": \"A\",\n",
        "    \"UGU\": \"C\", \"UGC\": \"C\",\n",
        "    \"GAU\": \"D\", \"GAC\": \"D\",\n",
        "    \"GAA\": \"E\", \"GAG\": \"E\",\n",
        "    \"UUU\": \"F\", \"UUC\": \"F\",\n",
        "    \"GGU\": \"G\", \"GGC\": \"G\", \"GGA\": \"G\", \"GGG\": \"G\",\n",
        "    \"CAU\": \"H\", \"CAC\": \"H\",\n",
        "    \"AUA\": \"I\", \"AUU\": \"I\", \"AUC\": \"I\",\n",
        "    \"AAA\": \"K\", \"AAG\": \"K\",\n",
        "    \"UUA\": \"L\", \"UUG\": \"L\", \"CUU\": \"L\", \"CUC\": \"L\", \"CUA\": \"L\", \"CUG\": \"L\",\n",
        "    \"AUG\": \"M\",\n",
        "    \"AAU\": \"N\", \"AAC\": \"N\",\n",
        "    \"CCU\": \"P\", \"CCC\": \"P\", \"CCA\": \"P\", \"CCG\": \"P\",\n",
        "    \"CAA\": \"Q\", \"CAG\": \"Q\",\n",
        "    \"CGU\": \"R\", \"CGC\": \"R\", \"CGA\": \"R\", \"CGG\": \"R\", \"AGA\": \"R\", \"AGG\": \"R\",\n",
        "    \"UCU\": \"S\", \"UCC\": \"S\", \"UCA\": \"S\", \"UCG\": \"S\", \"AGU\": \"S\", \"AGC\": \"S\",\n",
        "    \"ACU\": \"T\", \"ACC\": \"T\", \"ACA\": \"T\", \"ACG\": \"T\",\n",
        "    \"GUU\": \"V\", \"GUC\": \"V\", \"GUA\": \"V\", \"GUG\": \"V\",\n",
        "    \"UGG\": \"W\",\n",
        "    \"UAU\": \"Y\", \"UAC\": \"Y\",\n",
        "    \"UAA\": \"_\", \"UAG\": \"_\", \"UGA\": \"_\"\n",
        "}\n"
      ],
      "metadata": {
        "id": "PU6lzWHdK0RC"
      },
      "execution_count": 8,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "dna_random= ''.join([random.choice(nucleodites) for nuc in range(30)])\n",
        "print('\\n Check if my Data is a DNA Sequence \\n')\n",
        "print(dna_colors(validate_seg_dna(dna_random)))\n",
        "\n",
        "print('\\n Check the Number of Nucleodites I have in my DNA list \\n')\n",
        "dna_counts = count_nucleodites(dna_random)\n",
        "formatted_counts = f\"A: {dna_counts['A']}, T: {dna_counts['T']}, C: {dna_counts['C']}, G: {dna_counts['G']}\"\n",
        "print(dna_colors(formatted_counts))\n",
        "print(f' '.join([str(val) for key,val in count_nucleodites(dna_random).items()]))\n",
        "\n",
        "print('\\n Transcription of DNA to RNA \\n')\n",
        "print(dna_colors(transcription(dna_random)))\n",
        "\n",
        "print('\\n Implement my DNA  Reverse Complement \\n')\n",
        "print(f\"DNA + Reverse: \\n 5' {dna_colors(dna_random)} 3'\")\n",
        "print(f\"    {''.join(['|' for nuc in range(len(dna_random))])}\")\n",
        "print(f\"\\n\\n  {dna_colors(reverse_complement(dna_random))}  \\n\")\n",
        "\n",
        "print('\\n Implement my DNA complement \\n')\n",
        "print(f\"DNA + Complement: \\n 5' {dna_colors(dna_random)} 3'\")\n",
        "print(f\"    {''.join(['|' for nuc in range(len(dna_random))])}\")\n",
        "print(f\" 3' {dna_colors(dna_complement(dna_random))} 5'\\n\")\n",
        "\n",
        "#GC-rich regions may indicate greater compaction, greater genetic density,\n",
        "#greater stability and resistance to dematuration and may also indicate the origin of genes\n",
        "#in foreign genomes\n",
        "print('\\n Calculate my GC Content total and subsets \\n')\n",
        "print('\\nTotal GC Content \\n')\n",
        "print(gc_content(dna_random))\n",
        "print('\\nGC Content Subsets \\n')\n",
        "print(gc_content_subsets(dna_random,k=5))\n",
        "\n",
        "#Read information from a file in a specific format and perform the GC content calculation\n",
        "\n",
        "Fastadict=read_FASTA('fasta_samples.txt')\n",
        "\n",
        "DictGC={key:gc_content(value) for (key,value) in Fastadict.items()}\n",
        "\n",
        "maxGC = (max(DictGC, key=DictGC.get))\n",
        "print(f'{maxGC}\\n{DictGC[maxGC]}')\n",
        "\n",
        "print(f'\\n Amino Acid Sequence \\n')\n",
        "print(translate_dna(dna_random))\n",
        "print(f'\\n Frequency of a codon and returns the amino acid \\n')\n",
        "print(codon_freq(dna_random,'L'))\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "O3c1RtYWCvv4",
        "outputId": "1475493e-d9aa-4462-ffd8-dafb4bb0d50b"
      },
      "execution_count": 15,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            " Check if my Data is a DNA Sequence \n",
            "\n",
            "\u001b[94mC\u001b[94mC\u001b[93mG\u001b[94mC\u001b[92mA\u001b[93mG\u001b[94mC\u001b[91mT\u001b[91mT\u001b[91mT\u001b[92mA\u001b[91mT\u001b[92mA\u001b[91mT\u001b[92mA\u001b[92mA\u001b[92mA\u001b[91mT\u001b[92mA\u001b[94mC\u001b[91mT\u001b[92mA\u001b[92mA\u001b[91mT\u001b[92mA\u001b[93mG\u001b[93mG\u001b[91mT\u001b[92mA\u001b[93mG\u001b[0;0m\n",
            "\n",
            " Check the Number of Nucleodites I have in my DNA list \n",
            "\n",
            "\u001b[92mA\u001b[0;0m:\u001b[0;0m \u001b[0;0m1\u001b[0;0m1\u001b[0;0m,\u001b[0;0m \u001b[91mT\u001b[0;0m:\u001b[0;0m \u001b[0;0m9\u001b[0;0m,\u001b[0;0m \u001b[94mC\u001b[0;0m:\u001b[0;0m \u001b[0;0m5\u001b[0;0m,\u001b[0;0m \u001b[93mG\u001b[0;0m:\u001b[0;0m \u001b[0;0m5\u001b[0;0m\n",
            "11 9 5 5\n",
            "\n",
            " Transcription of DNA to RNA \n",
            "\n",
            "\u001b[94mC\u001b[94mC\u001b[93mG\u001b[94mC\u001b[92mA\u001b[93mG\u001b[94mC\u001b[91mU\u001b[91mU\u001b[91mU\u001b[92mA\u001b[91mU\u001b[92mA\u001b[91mU\u001b[92mA\u001b[92mA\u001b[92mA\u001b[91mU\u001b[92mA\u001b[94mC\u001b[91mU\u001b[92mA\u001b[92mA\u001b[91mU\u001b[92mA\u001b[93mG\u001b[93mG\u001b[91mU\u001b[92mA\u001b[93mG\u001b[0;0m\n",
            "\n",
            " Implement my DNA  Reverse Complement \n",
            "\n",
            "DNA + Reverse: \n",
            " 5' \u001b[94mC\u001b[94mC\u001b[93mG\u001b[94mC\u001b[92mA\u001b[93mG\u001b[94mC\u001b[91mT\u001b[91mT\u001b[91mT\u001b[92mA\u001b[91mT\u001b[92mA\u001b[91mT\u001b[92mA\u001b[92mA\u001b[92mA\u001b[91mT\u001b[92mA\u001b[94mC\u001b[91mT\u001b[92mA\u001b[92mA\u001b[91mT\u001b[92mA\u001b[93mG\u001b[93mG\u001b[91mT\u001b[92mA\u001b[93mG\u001b[0;0m 3'\n",
            "    ||||||||||||||||||||||||||||||\n",
            "\n",
            "\n",
            "  \u001b[94mC\u001b[91mT\u001b[92mA\u001b[94mC\u001b[94mC\u001b[91mT\u001b[92mA\u001b[91mT\u001b[91mT\u001b[92mA\u001b[93mG\u001b[91mT\u001b[92mA\u001b[91mT\u001b[91mT\u001b[91mT\u001b[92mA\u001b[91mT\u001b[92mA\u001b[91mT\u001b[92mA\u001b[92mA\u001b[92mA\u001b[93mG\u001b[94mC\u001b[91mT\u001b[93mG\u001b[94mC\u001b[93mG\u001b[93mG\u001b[0;0m  \n",
            "\n",
            "\n",
            " Implement my DNA complement \n",
            "\n",
            "DNA + Complement: \n",
            " 5' \u001b[94mC\u001b[94mC\u001b[93mG\u001b[94mC\u001b[92mA\u001b[93mG\u001b[94mC\u001b[91mT\u001b[91mT\u001b[91mT\u001b[92mA\u001b[91mT\u001b[92mA\u001b[91mT\u001b[92mA\u001b[92mA\u001b[92mA\u001b[91mT\u001b[92mA\u001b[94mC\u001b[91mT\u001b[92mA\u001b[92mA\u001b[91mT\u001b[92mA\u001b[93mG\u001b[93mG\u001b[91mT\u001b[92mA\u001b[93mG\u001b[0;0m 3'\n",
            "    ||||||||||||||||||||||||||||||\n",
            " 3' \u001b[93mG\u001b[93mG\u001b[94mC\u001b[93mG\u001b[91mT\u001b[94mC\u001b[93mG\u001b[92mA\u001b[92mA\u001b[92mA\u001b[91mT\u001b[92mA\u001b[91mT\u001b[92mA\u001b[91mT\u001b[91mT\u001b[91mT\u001b[92mA\u001b[91mT\u001b[93mG\u001b[92mA\u001b[91mT\u001b[91mT\u001b[92mA\u001b[91mT\u001b[94mC\u001b[94mC\u001b[92mA\u001b[91mT\u001b[94mC\u001b[0;0m 5'\n",
            "\n",
            "\n",
            " Calculate my GC Content total and subsets \n",
            "\n",
            "\n",
            "Total GC Content \n",
            "\n",
            "33.333333\n",
            "\n",
            "GC Content Subsets \n",
            "\n",
            "[80.0, 40.0, 0.0, 20.0, 0.0, 60.0]\n",
            ">Rosalind_0808\n",
            "60.91954\n",
            "\n",
            " Amino Acid Sequence \n",
            "\n",
            "['P', 'Q', 'L', 'Y', 'I', 'N', 'T', 'N', 'R', '_']\n",
            "\n",
            " Frequency of a codon and returns the amino acid \n",
            "\n",
            "{'CTT': 1.0}\n"
          ]
        }
      ]
    }
  ]
}