{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyOZjERTjyGz/DPD1dvko+9g"
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "krIK_AtSBq55"
      },
      "outputs": [],
      "source": [
        "from typing_extensions import Counter\n",
        "import collections\n",
        "import random\n",
        "\n",
        "nucleodites =['A','T','C','G']\n",
        "reverse_nucleodites = {'A':'T','T':'A','G':'C','C':'G'}\n",
        "\n",
        "def validate_seg_dna(dna):\n",
        "    \"\"\"Validate if a sequence is a DNA sequence\"\"\"\n",
        "    verification = dna.upper()\n",
        "    for nuc in verification:\n",
        "        if nuc not in nucleodites:\n",
        "            return False\n",
        "    return verification\n",
        "\n",
        "def count_nucleodites(dna):\n",
        "    \"\"\"Count the number of nucleodites in a DNA sequence\"\"\"\n",
        "    nucleodites_count = {'A':0,'T':0,'C':0,'G':0}\n",
        "    dna = dna.upper()\n",
        "    for nuc in dna:\n",
        "        nucleodites_count[nuc] += 1\n",
        "    return nucleodites_count\n",
        "\n",
        "def transcription(dna):\n",
        "    \"\"\"Return a list of nucleodites to complete the dna\"\"\"\n",
        "    dna = dna.upper()\n",
        "    return dna.replace('T','U')\n",
        "\n",
        "def reverse_complement(dna):\n",
        "    \"\"\"Return a list of nucleodites to complete the dna\"\"\"\n",
        "    #return ''.join([reverse_nucleodites[nuc] for nuc in dna])[::-1]\n",
        "    dna = dna.upper()\n",
        "    mapping=str(dna).maketrans('ATCG','TAGC')\n",
        "    return dna.translate(mapping)[::-1]\n",
        "\n",
        "\n",
        "def dna_complement(dna):\n",
        "    \"\"\"Return a list of nucleodites reverse of a list of nucleodites\"\"\"\n",
        "    return ''.join([reverse_nucleodites[nuc.upper()] for nuc in dna])\n",
        "\n",
        "def dna_colors(dna):\n",
        "    \"\"\"Define a color for each nucleotide in a DNA sequence\"\"\"\n",
        "    #dna = dna.upper()\n",
        "    dnacolors = {\n",
        "        'A': '\\033[92m',\n",
        "        'C': '\\033[94m',\n",
        "        'G': '\\033[93m',\n",
        "        'T': '\\033[91m',\n",
        "        'U': '\\033[91m',\n",
        "        'reset': '\\033[0;0m'\n",
        "    }\n",
        "\n",
        "    variavel = \"\"\n",
        "\n",
        "    for nuc in dna:\n",
        "        if nuc in dnacolors:\n",
        "            variavel += dnacolors[nuc] + nuc\n",
        "        else:\n",
        "            variavel += dnacolors['reset'] + nuc\n",
        "\n",
        "    return variavel + '\\033[0;0m'\n",
        "\n",
        "def gc_content(dna):\n",
        "  \"\"\"GC Content in a DNA/RNA sequence\"\"\"\n",
        "  dna = dna.upper()\n",
        "  if len(dna) == 0:\n",
        "    return 0\n",
        "  return round((dna.count('C') + dna.count('G')) / len(dna) * 100,6)\n",
        "\n",
        "def gc_content_subsets(dna,k=20):\n",
        "  \"\"\"GC Content in a DNA/RNA sub-sequence length k. k=20 by default\"\"\"\n",
        "  dna = dna.upper()\n",
        "  props=[]\n",
        "  for i in range(0,len(dna)-k+1,k):\n",
        "    subsets=dna[i:i+k]\n",
        "    props.append(gc_content(subsets))\n",
        "  return props\n",
        "\n",
        "\n",
        "def readTextFile(filePath):\n",
        "    \"\"\"Read a text file and return a string\"\"\"\n",
        "    with open(filePath, 'r') as f:\n",
        "        return \"\".join([l.strip() for l in f.readlines()])\n",
        "\n",
        "\n",
        "def writeTextFile(filePath, dna, mode='w'):\n",
        "    \"\"\"Write a DNA sequence to a file\"\"\"\n",
        "    with open(filePath, mode) as f:\n",
        "        f.write(dna + '\\n')\n",
        "\n",
        "\n",
        "def read_FASTA(filePath):\n",
        "    \"\"\"Read a FASTA file and return a dictionary with the sequence\"\"\"\n",
        "    with open(filePath, 'r') as f:\n",
        "        FASTAFile = [l.strip() for l in f.readlines()]\n",
        "\n",
        "    FASTADict = {}\n",
        "    FASTALabel = \"\"\n",
        "\n",
        "    for line in FASTAFile:\n",
        "        if '>' in line:\n",
        "            FASTALabel = line\n",
        "            FASTADict[FASTALabel] = \"\"\n",
        "        else:\n",
        "            FASTADict[FASTALabel] += line\n",
        "\n",
        "    return FASTADict\n",
        "\n",
        "# List the amoni acids\n",
        "def translate_dna(dna,position_initial=0):\n",
        "    \"\"\"Translates a DNA sequence into an aminoacid sequence\"\"\"\n",
        "    #Check if dna is a string\n",
        "    if isinstance(dna, str):\n",
        "        return [DNA_Codons[dna[pos:pos + 3]] for pos in range(position_initial, len(dna)-2, 3)]\n",
        "    #If dna is not a string (e.g., a list), iterate over its elements and translate them\n",
        "    elif isinstance(dna, list):\n",
        "        translated_dna = []\n",
        "        for item in dna:\n",
        "            #Check if each element is a string\n",
        "            if isinstance(item, str):\n",
        "                #Use the original translate_dna function on strings\n",
        "                translated_dna.extend(translate_dna(item, position_initial))\n",
        "            else:\n",
        "                #Handle cases for non-string elements if necessary\n",
        "                pass\n",
        "        return translated_dna\n",
        "    else:\n",
        "        return []\n",
        "\n",
        "#Frequency of a codon and returns the amino acid\n",
        "def codon_freq(dna,animoacid):\n",
        "  \"\"\"Provides the frequency of each codon encoding a given aminoacid in a DNA sequence\"\"\"\n",
        "  dna = dna.upper()\n",
        "  lista=[]\n",
        "  for i in range(0,len(dna)-2,3):\n",
        "    if DNA_Codons[dna[i:i+3]]==animoacid:\n",
        "      lista.append(dna[i:i+3])\n",
        "  freqdic=dict(Counter(lista))\n",
        "  total_freq=sum(freqdic.values())\n",
        "  for dna in freqdic:\n",
        "    freqdic[dna]=round(freqdic[dna]/total_freq,2)\n",
        "  return freqdic\n",
        "\n",
        "#In molecular biology, a reading frame is a specific choice out of the possible ways to read the sequence of nucleotides in a nucleic acid (DNA or RNA)\n",
        "# molecule as a sequence of triplets. Where these triplets equate to amino acids or stop signals during translation, they are called codons.\n",
        "def reading_frames(dna):\n",
        "  \"\"\"Generate reading frames of a dna sequence, including reverse complement\"\"\"\n",
        "  frames=[]\n",
        "  frames.append(translate_dna(dna,0))\n",
        "  frames.append(translate_dna(dna,1))\n",
        "  frames.append(translate_dna(dna,2))\n",
        "  frames.append(translate_dna(reverse_complement(dna),0))\n",
        "  frames.append(translate_dna(reverse_complement(dna),1))\n",
        "  frames.append(translate_dna(reverse_complement(dna),2))\n",
        "  return frames\n",
        "\n",
        "\n",
        "def proteins(dna):\n",
        "  \"\"\"Compute all possible proteins and return a list of possible proteins\"\"\"\n",
        "  proteinas=[]\n",
        "  proteina_atual=[]\n",
        "  for amino in dna:\n",
        "    if amino ==\"_\":\n",
        "      if proteina_atual:\n",
        "        for p in proteina_atual:\n",
        "          proteinas.append(''.join(p))\n",
        "          proteina_atual=[]\n",
        "    else:\n",
        "      if amino == \"M\":\n",
        "        proteina_atual.append(\"\")\n",
        "      for i in range(len(proteina_atual)):\n",
        "        proteina_atual[i]+=amino\n",
        "  return proteinas\n",
        "\n",
        "def frames_proteins(dna,inicio=0,final=0,order='False'):\n",
        "   \"\"\"Compute all possible proteins for all reading frames\"\"\"\n",
        "   \"\"\"Protine Search DB: https://www.ncbi.nlm.nih.gov/nuccore/NM_001185097.2\"\"\"\n",
        "   \"\"\"API can be used to pull protein info\"\"\"\n",
        "   if final> inicio:\n",
        "    frames=reading_frames(dna[inicio:final])\n",
        "   else:\n",
        "    frames = reading_frames(dna)\n",
        "\n",
        "   find_prot=[]\n",
        "   for frame in frames:\n",
        "    proteinas=proteins(frame)\n",
        "    for prot in proteinas:\n",
        "      find_prot.append(prot)\n",
        "   if order=='True':\n",
        "     find_prot.sort(key=len)\n",
        "   return find_prot\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "NUCLEOTIDE_BASE = {\n",
        "    \"DNA\": [\"A\", \"T\", \"C\", \"G\"],\n",
        "    \"RNA\": [\"A\", \"U\", \"C\", \"G\"]\n",
        "}\n",
        "\n",
        "DNA_Codons = {\n",
        "    # 'M' - START, '_' - STOP\n",
        "    \"GCT\": \"A\", \"GCC\": \"A\", \"GCA\": \"A\", \"GCG\": \"A\",\n",
        "    \"TGT\": \"C\", \"TGC\": \"C\",\n",
        "    \"GAT\": \"D\", \"GAC\": \"D\",\n",
        "    \"GAA\": \"E\", \"GAG\": \"E\",\n",
        "    \"TTT\": \"F\", \"TTC\": \"F\",\n",
        "    \"GGT\": \"G\", \"GGC\": \"G\", \"GGA\": \"G\", \"GGG\": \"G\",\n",
        "    \"CAT\": \"H\", \"CAC\": \"H\",\n",
        "    \"ATA\": \"I\", \"ATT\": \"I\", \"ATC\": \"I\",\n",
        "    \"AAA\": \"K\", \"AAG\": \"K\",\n",
        "    \"TTA\": \"L\", \"TTG\": \"L\", \"CTT\": \"L\", \"CTC\": \"L\", \"CTA\": \"L\", \"CTG\": \"L\",\n",
        "    \"ATG\": \"M\",\n",
        "    \"AAT\": \"N\", \"AAC\": \"N\",\n",
        "    \"CCT\": \"P\", \"CCC\": \"P\", \"CCA\": \"P\", \"CCG\": \"P\",\n",
        "    \"CAA\": \"Q\", \"CAG\": \"Q\",\n",
        "    \"CGT\": \"R\", \"CGC\": \"R\", \"CGA\": \"R\", \"CGG\": \"R\", \"AGA\": \"R\", \"AGG\": \"R\",\n",
        "    \"TCT\": \"S\", \"TCC\": \"S\", \"TCA\": \"S\", \"TCG\": \"S\", \"AGT\": \"S\", \"AGC\": \"S\",\n",
        "    \"ACT\": \"T\", \"ACC\": \"T\", \"ACA\": \"T\", \"ACG\": \"T\",\n",
        "    \"GTT\": \"V\", \"GTC\": \"V\", \"GTA\": \"V\", \"GTG\": \"V\",\n",
        "    \"TGG\": \"W\",\n",
        "    \"TAT\": \"Y\", \"TAC\": \"Y\",\n",
        "    \"TAA\": \"_\", \"TAG\": \"_\", \"TGA\": \"_\"\n",
        "}\n",
        "\n",
        "RNA_Codons = {\n",
        "    # 'M' - START, '_' - STOP\n",
        "    \"GCU\": \"A\", \"GCC\": \"A\", \"GCA\": \"A\", \"GCG\": \"A\",\n",
        "    \"UGU\": \"C\", \"UGC\": \"C\",\n",
        "    \"GAU\": \"D\", \"GAC\": \"D\",\n",
        "    \"GAA\": \"E\", \"GAG\": \"E\",\n",
        "    \"UUU\": \"F\", \"UUC\": \"F\",\n",
        "    \"GGU\": \"G\", \"GGC\": \"G\", \"GGA\": \"G\", \"GGG\": \"G\",\n",
        "    \"CAU\": \"H\", \"CAC\": \"H\",\n",
        "    \"AUA\": \"I\", \"AUU\": \"I\", \"AUC\": \"I\",\n",
        "    \"AAA\": \"K\", \"AAG\": \"K\",\n",
        "    \"UUA\": \"L\", \"UUG\": \"L\", \"CUU\": \"L\", \"CUC\": \"L\", \"CUA\": \"L\", \"CUG\": \"L\",\n",
        "    \"AUG\": \"M\",\n",
        "    \"AAU\": \"N\", \"AAC\": \"N\",\n",
        "    \"CCU\": \"P\", \"CCC\": \"P\", \"CCA\": \"P\", \"CCG\": \"P\",\n",
        "    \"CAA\": \"Q\", \"CAG\": \"Q\",\n",
        "    \"CGU\": \"R\", \"CGC\": \"R\", \"CGA\": \"R\", \"CGG\": \"R\", \"AGA\": \"R\", \"AGG\": \"R\",\n",
        "    \"UCU\": \"S\", \"UCC\": \"S\", \"UCA\": \"S\", \"UCG\": \"S\", \"AGU\": \"S\", \"AGC\": \"S\",\n",
        "    \"ACU\": \"T\", \"ACC\": \"T\", \"ACA\": \"T\", \"ACG\": \"T\",\n",
        "    \"GUU\": \"V\", \"GUC\": \"V\", \"GUA\": \"V\", \"GUG\": \"V\",\n",
        "    \"UGG\": \"W\",\n",
        "    \"UAU\": \"Y\", \"UAC\": \"Y\",\n",
        "    \"UAA\": \"_\", \"UAG\": \"_\", \"UGA\": \"_\"\n",
        "}\n"
      ],
      "metadata": {
        "id": "PU6lzWHdK0RC"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#Read information from a file in a specific format and perform the GC content calculation\n",
        "Fastadict=read_FASTA('fasta_samples.txt')\n",
        "first_sequence_key = list(Fastadict.keys())[0]\n",
        "first_sequence_value = Fastadict[first_sequence_key]\n",
        "dna_random= first_sequence_value\n",
        "print('\\n Check if my Data is a DNA Sequence \\n')\n",
        "print(dna_colors(validate_seg_dna(dna_random)))\n",
        "\n",
        "print('\\n Check the Number of Nucleodites I have in my DNA list \\n')\n",
        "dna_counts = count_nucleodites(dna_random)\n",
        "formatted_counts = f\"A: {dna_counts['A']}, T: {dna_counts['T']}, C: {dna_counts['C']}, G: {dna_counts['G']}\"\n",
        "print(dna_colors(formatted_counts))\n",
        "print(f' '.join([str(val) for key,val in count_nucleodites(dna_random).items()]))\n",
        "\n",
        "print('\\n Transcription of DNA to RNA \\n')\n",
        "print(dna_colors(transcription(dna_random)))\n",
        "\n",
        "print('\\n Implement my DNA  Reverse Complement \\n')\n",
        "print(f\"DNA + Reverse: \\n 5' {dna_colors(dna_random)} 3'\")\n",
        "print(f\"    {''.join(['|' for nuc in range(len(dna_random))])}\")\n",
        "print(f\"\\n\\n  {dna_colors(reverse_complement(dna_random))}  \\n\")\n",
        "\n",
        "print('\\n Implement my DNA complement \\n')\n",
        "print(f\"DNA + Complement: \\n 5' {dna_colors(dna_random)} 3'\")\n",
        "print(f\"    {''.join(['|' for nuc in range(len(dna_random))])}\")\n",
        "print(f\" 3' {dna_colors(dna_complement(dna_random))} 5'\\n\")\n",
        "\n",
        "#GC-rich regions may indicate greater compaction, greater genetic density,\n",
        "#greater stability and resistance to dematuration and may also indicate the origin of genes\n",
        "#in foreign genomes\n",
        "print('\\n Calculate my GC Content total and subsets \\n')\n",
        "print('\\nTotal GC Content \\n')\n",
        "print(gc_content(dna_random))\n",
        "print('\\nGC Content Subsets \\n')\n",
        "print(gc_content_subsets(dna_random,k=5))\n",
        "\n",
        "DictGC={key:gc_content(value) for (key,value) in Fastadict.items()}\n",
        "\n",
        "maxGC = (max(DictGC, key=DictGC.get))\n",
        "print(f'{maxGC}\\n{DictGC[maxGC]}')\n",
        "\n",
        "print(f'\\n Amino Acid Sequence \\n')\n",
        "print(translate_dna(dna_random))\n",
        "print(f'\\n Frequency of a codon and returns the amino acid \\n')\n",
        "print(codon_freq(dna_random,'P'))\n",
        "\n",
        "print(f'\\n Reading Frames \\n')\n",
        "for reading in reading_frames(dna_random):\n",
        "  print(reading)\n",
        "print(f'\\n Frames Proteins \\n')\n",
        "for protein in frames_proteins(dna_random):\n",
        "  print(protein)\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "O3c1RtYWCvv4",
        "outputId": "2a38879b-c5b5-471c-b7fd-2dfc67a31666"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            " Check if my Data is a DNA Sequence \n",
            "\n",
            "\u001b[94mC\u001b[94mC\u001b[91mT\u001b[93mG\u001b[94mC\u001b[93mG\u001b[93mG\u001b[92mA\u001b[92mA\u001b[93mG\u001b[92mA\u001b[91mT\u001b[94mC\u001b[93mG\u001b[93mG\u001b[94mC\u001b[92mA\u001b[94mC\u001b[91mT\u001b[92mA\u001b[93mG\u001b[92mA\u001b[92mA\u001b[91mT\u001b[92mA\u001b[93mG\u001b[94mC\u001b[94mC\u001b[92mA\u001b[93mG\u001b[92mA\u001b[92mA\u001b[94mC\u001b[94mC\u001b[93mG\u001b[91mT\u001b[91mT\u001b[91mT\u001b[94mC\u001b[91mT\u001b[94mC\u001b[91mT\u001b[93mG\u001b[92mA\u001b[93mG\u001b[93mG\u001b[94mC\u001b[91mT\u001b[91mT\u001b[94mC\u001b[94mC\u001b[93mG\u001b[93mG\u001b[94mC\u001b[94mC\u001b[91mT\u001b[91mT\u001b[94mC\u001b[94mC\u001b[94mC\u001b[91mT\u001b[94mC\u001b[94mC\u001b[94mC\u001b[92mA\u001b[94mC\u001b[91mT\u001b[92mA\u001b[92mA\u001b[91mT\u001b[92mA\u001b[92mA\u001b[91mT\u001b[91mT\u001b[94mC\u001b[91mT\u001b[93mG\u001b[92mA\u001b[93mG\u001b[93mG\u001b[0;0m\n",
            "\n",
            " Check the Number of Nucleodites I have in my DNA list \n",
            "\n",
            "\u001b[92mA\u001b[0;0m:\u001b[0;0m \u001b[0;0m1\u001b[0;0m8\u001b[0;0m,\u001b[0;0m \u001b[91mT\u001b[0;0m:\u001b[0;0m \u001b[0;0m1\u001b[0;0m9\u001b[0;0m,\u001b[0;0m \u001b[94mC\u001b[0;0m:\u001b[0;0m \u001b[0;0m2\u001b[0;0m5\u001b[0;0m,\u001b[0;0m \u001b[93mG\u001b[0;0m:\u001b[0;0m \u001b[0;0m1\u001b[0;0m8\u001b[0;0m\n",
            "18 19 25 18\n",
            "\n",
            " Transcription of DNA to RNA \n",
            "\n",
            "\u001b[94mC\u001b[94mC\u001b[91mU\u001b[93mG\u001b[94mC\u001b[93mG\u001b[93mG\u001b[92mA\u001b[92mA\u001b[93mG\u001b[92mA\u001b[91mU\u001b[94mC\u001b[93mG\u001b[93mG\u001b[94mC\u001b[92mA\u001b[94mC\u001b[91mU\u001b[92mA\u001b[93mG\u001b[92mA\u001b[92mA\u001b[91mU\u001b[92mA\u001b[93mG\u001b[94mC\u001b[94mC\u001b[92mA\u001b[93mG\u001b[92mA\u001b[92mA\u001b[94mC\u001b[94mC\u001b[93mG\u001b[91mU\u001b[91mU\u001b[91mU\u001b[94mC\u001b[91mU\u001b[94mC\u001b[91mU\u001b[93mG\u001b[92mA\u001b[93mG\u001b[93mG\u001b[94mC\u001b[91mU\u001b[91mU\u001b[94mC\u001b[94mC\u001b[93mG\u001b[93mG\u001b[94mC\u001b[94mC\u001b[91mU\u001b[91mU\u001b[94mC\u001b[94mC\u001b[94mC\u001b[91mU\u001b[94mC\u001b[94mC\u001b[94mC\u001b[92mA\u001b[94mC\u001b[91mU\u001b[92mA\u001b[92mA\u001b[91mU\u001b[92mA\u001b[92mA\u001b[91mU\u001b[91mU\u001b[94mC\u001b[91mU\u001b[93mG\u001b[92mA\u001b[93mG\u001b[93mG\u001b[0;0m\n",
            "\n",
            " Implement my DNA  Reverse Complement \n",
            "\n",
            "DNA + Reverse: \n",
            " 5' \u001b[94mC\u001b[94mC\u001b[91mT\u001b[93mG\u001b[94mC\u001b[93mG\u001b[93mG\u001b[92mA\u001b[92mA\u001b[93mG\u001b[92mA\u001b[91mT\u001b[94mC\u001b[93mG\u001b[93mG\u001b[94mC\u001b[92mA\u001b[94mC\u001b[91mT\u001b[92mA\u001b[93mG\u001b[92mA\u001b[92mA\u001b[91mT\u001b[92mA\u001b[93mG\u001b[94mC\u001b[94mC\u001b[92mA\u001b[93mG\u001b[92mA\u001b[92mA\u001b[94mC\u001b[94mC\u001b[93mG\u001b[91mT\u001b[91mT\u001b[91mT\u001b[94mC\u001b[91mT\u001b[94mC\u001b[91mT\u001b[93mG\u001b[92mA\u001b[93mG\u001b[93mG\u001b[94mC\u001b[91mT\u001b[91mT\u001b[94mC\u001b[94mC\u001b[93mG\u001b[93mG\u001b[94mC\u001b[94mC\u001b[91mT\u001b[91mT\u001b[94mC\u001b[94mC\u001b[94mC\u001b[91mT\u001b[94mC\u001b[94mC\u001b[94mC\u001b[92mA\u001b[94mC\u001b[91mT\u001b[92mA\u001b[92mA\u001b[91mT\u001b[92mA\u001b[92mA\u001b[91mT\u001b[91mT\u001b[94mC\u001b[91mT\u001b[93mG\u001b[92mA\u001b[93mG\u001b[93mG\u001b[0;0m 3'\n",
            "    ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
            "\n",
            "\n",
            "  \u001b[94mC\u001b[94mC\u001b[91mT\u001b[94mC\u001b[92mA\u001b[93mG\u001b[92mA\u001b[92mA\u001b[91mT\u001b[91mT\u001b[92mA\u001b[91mT\u001b[91mT\u001b[92mA\u001b[93mG\u001b[91mT\u001b[93mG\u001b[93mG\u001b[93mG\u001b[92mA\u001b[93mG\u001b[93mG\u001b[93mG\u001b[92mA\u001b[92mA\u001b[93mG\u001b[93mG\u001b[94mC\u001b[94mC\u001b[93mG\u001b[93mG\u001b[92mA\u001b[92mA\u001b[93mG\u001b[94mC\u001b[94mC\u001b[91mT\u001b[94mC\u001b[92mA\u001b[93mG\u001b[92mA\u001b[93mG\u001b[92mA\u001b[92mA\u001b[92mA\u001b[94mC\u001b[93mG\u001b[93mG\u001b[91mT\u001b[91mT\u001b[94mC\u001b[91mT\u001b[93mG\u001b[93mG\u001b[94mC\u001b[91mT\u001b[92mA\u001b[91mT\u001b[91mT\u001b[94mC\u001b[91mT\u001b[92mA\u001b[93mG\u001b[91mT\u001b[93mG\u001b[94mC\u001b[94mC\u001b[93mG\u001b[92mA\u001b[91mT\u001b[94mC\u001b[91mT\u001b[91mT\u001b[94mC\u001b[94mC\u001b[93mG\u001b[94mC\u001b[92mA\u001b[93mG\u001b[93mG\u001b[0;0m  \n",
            "\n",
            "\n",
            " Implement my DNA complement \n",
            "\n",
            "DNA + Complement: \n",
            " 5' \u001b[94mC\u001b[94mC\u001b[91mT\u001b[93mG\u001b[94mC\u001b[93mG\u001b[93mG\u001b[92mA\u001b[92mA\u001b[93mG\u001b[92mA\u001b[91mT\u001b[94mC\u001b[93mG\u001b[93mG\u001b[94mC\u001b[92mA\u001b[94mC\u001b[91mT\u001b[92mA\u001b[93mG\u001b[92mA\u001b[92mA\u001b[91mT\u001b[92mA\u001b[93mG\u001b[94mC\u001b[94mC\u001b[92mA\u001b[93mG\u001b[92mA\u001b[92mA\u001b[94mC\u001b[94mC\u001b[93mG\u001b[91mT\u001b[91mT\u001b[91mT\u001b[94mC\u001b[91mT\u001b[94mC\u001b[91mT\u001b[93mG\u001b[92mA\u001b[93mG\u001b[93mG\u001b[94mC\u001b[91mT\u001b[91mT\u001b[94mC\u001b[94mC\u001b[93mG\u001b[93mG\u001b[94mC\u001b[94mC\u001b[91mT\u001b[91mT\u001b[94mC\u001b[94mC\u001b[94mC\u001b[91mT\u001b[94mC\u001b[94mC\u001b[94mC\u001b[92mA\u001b[94mC\u001b[91mT\u001b[92mA\u001b[92mA\u001b[91mT\u001b[92mA\u001b[92mA\u001b[91mT\u001b[91mT\u001b[94mC\u001b[91mT\u001b[93mG\u001b[92mA\u001b[93mG\u001b[93mG\u001b[0;0m 3'\n",
            "    ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
            " 3' \u001b[93mG\u001b[93mG\u001b[92mA\u001b[94mC\u001b[93mG\u001b[94mC\u001b[94mC\u001b[91mT\u001b[91mT\u001b[94mC\u001b[91mT\u001b[92mA\u001b[93mG\u001b[94mC\u001b[94mC\u001b[93mG\u001b[91mT\u001b[93mG\u001b[92mA\u001b[91mT\u001b[94mC\u001b[91mT\u001b[91mT\u001b[92mA\u001b[91mT\u001b[94mC\u001b[93mG\u001b[93mG\u001b[91mT\u001b[94mC\u001b[91mT\u001b[91mT\u001b[93mG\u001b[93mG\u001b[94mC\u001b[92mA\u001b[92mA\u001b[92mA\u001b[93mG\u001b[92mA\u001b[93mG\u001b[92mA\u001b[94mC\u001b[91mT\u001b[94mC\u001b[94mC\u001b[93mG\u001b[92mA\u001b[92mA\u001b[93mG\u001b[93mG\u001b[94mC\u001b[94mC\u001b[93mG\u001b[93mG\u001b[92mA\u001b[92mA\u001b[93mG\u001b[93mG\u001b[93mG\u001b[92mA\u001b[93mG\u001b[93mG\u001b[93mG\u001b[91mT\u001b[93mG\u001b[92mA\u001b[91mT\u001b[91mT\u001b[92mA\u001b[91mT\u001b[91mT\u001b[92mA\u001b[92mA\u001b[93mG\u001b[92mA\u001b[94mC\u001b[91mT\u001b[94mC\u001b[94mC\u001b[0;0m 5'\n",
            "\n",
            "\n",
            " Calculate my GC Content total and subsets \n",
            "\n",
            "\n",
            "Total GC Content \n",
            "\n",
            "53.75\n",
            "\n",
            "GC Content Subsets \n",
            "\n",
            "[80.0, 60.0, 60.0, 40.0, 20.0, 80.0, 60.0, 20.0, 60.0, 60.0, 100.0, 60.0, 60.0, 20.0, 20.0, 60.0]\n",
            ">Rosalind_0808\n",
            "60.91954\n",
            "\n",
            " Amino Acid Sequence \n",
            "\n",
            "['P', 'A', 'E', 'D', 'R', 'H', '_', 'N', 'S', 'Q', 'N', 'R', 'F', 'S', 'E', 'A', 'S', 'G', 'L', 'P', 'S', 'H', '_', '_', 'F', '_']\n",
            "\n",
            " Frequency of a codon and returns the amino acid \n",
            "\n",
            "{'CCT': 0.5, 'CCC': 0.5}\n",
            "\n",
            " Reading Frames \n",
            "\n",
            "['P', 'A', 'E', 'D', 'R', 'H', '_', 'N', 'S', 'Q', 'N', 'R', 'F', 'S', 'E', 'A', 'S', 'G', 'L', 'P', 'S', 'H', '_', '_', 'F', '_']\n",
            "['L', 'R', 'K', 'I', 'G', 'T', 'R', 'I', 'A', 'R', 'T', 'V', 'S', 'L', 'R', 'L', 'P', 'A', 'F', 'P', 'P', 'T', 'N', 'N', 'S', 'E']\n",
            "['C', 'G', 'R', 'S', 'A', 'L', 'E', '_', 'P', 'E', 'P', 'F', 'L', '_', 'G', 'F', 'R', 'P', 'S', 'L', 'P', 'L', 'I', 'I', 'L', 'R']\n",
            "['P', 'Q', 'N', 'Y', '_', 'W', 'E', 'G', 'R', 'P', 'E', 'A', 'S', 'E', 'K', 'R', 'F', 'W', 'L', 'F', '_', 'C', 'R', 'S', 'S', 'A']\n",
            "['L', 'R', 'I', 'I', 'S', 'G', 'R', 'E', 'G', 'R', 'K', 'P', 'Q', 'R', 'N', 'G', 'S', 'G', 'Y', 'S', 'S', 'A', 'D', 'L', 'P', 'Q']\n",
            "['S', 'E', 'L', 'L', 'V', 'G', 'G', 'K', 'A', 'G', 'S', 'L', 'R', 'E', 'T', 'V', 'L', 'A', 'I', 'L', 'V', 'P', 'I', 'F', 'R', 'R']\n",
            "\n",
            " Frames Proteins \n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#Applying the functions in the object-oriented model to create a package later"
      ],
      "metadata": {
        "id": "aAorVG0WK9ZW"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "from typing_extensions import Counter\n",
        "import collections\n",
        "\n",
        "\n",
        "class bioinformatics_funcions :\n",
        "  def __init__(self,dna='ATCG', seq_type='DNA',label='No Label'):\n",
        "    self.dna = dna\n",
        "    self.label = label\n",
        "    self.seq_type_ = seq_type\n",
        "    self.is_valid = self.validate_seg_dna()\n",
        "    assert self.is_valid, f\"Invalid {self.seq_type_} sequence\"\n",
        "\n",
        "  def dna_colors(self):\n",
        "      \"\"\"Define a color for each nucleotide in a DNA sequence\"\"\"\n",
        "      dna = self.dna.upper()\n",
        "      dnacolors = {\n",
        "          'A': '\\033[92m',\n",
        "          'C': '\\033[94m',\n",
        "          'G': '\\033[93m',\n",
        "          'T': '\\033[91m',\n",
        "          'U': '\\033[91m',\n",
        "          'reset': '\\033[0;0m'\n",
        "      }\n",
        "      variavel = \"\"\n",
        "\n",
        "      for nuc in dna:\n",
        "          if nuc in dnacolors:\n",
        "              variavel += dnacolors[nuc] + nuc\n",
        "          else:\n",
        "              variavel += dnacolors['reset'] + nuc\n",
        "\n",
        "      return variavel + '\\033[0;0m'\n",
        "\n",
        "  def values(self):\n",
        "        \"\"\"Returns a list of the values of the object's attributes.\"\"\"\n",
        "        return [f\"DNA: \\n 5' {self.dna} 3' \\n\"+f\"    {''.join(['|' for nuc in range(len(self.dna))])}\", self.label, self.seq_type_]\n",
        "  def validate_seg_dna(self):\n",
        "    \"\"\"Validate if a sequence is a DNA sequence\"\"\"\n",
        "    return set(NUCLEOTIDE_BASE[self.seq_type_]).issuperset(self.dna)\n",
        "\n",
        "  def seq_type(self):\n",
        "        \"\"\"Returns sequence type\"\"\"\n",
        "        return self.seq_type_\n",
        "\n",
        "  def seq_info(self):\n",
        "        \"\"\"Returns 4 strings. Full sequence information\"\"\"\n",
        "        return f\"[Label]: {self.label}\\n[Sequence]: {self.dna}\\n[Type]: {self.seq_type_}\\n[Length]: {len(self.dna)}\"\n",
        "\n",
        "\n",
        "\n",
        "  def count_nucleodites(self):\n",
        "    \"\"\"Count the number of nucleodites in a DNA sequence\"\"\"\n",
        "    if self.seq_type_ == \"DNA\":\n",
        "      nucleodites_count = {'A': 0, 'T': 0, 'C': 0, 'G': 0}\n",
        "      dna = self.dna.upper()\n",
        "      for nuc in dna:\n",
        "          nucleodites_count[nuc] += 1\n",
        "      # Format the output string using the nucleotide counts\n",
        "      #formatted_counts = f\"A: {nucleodites_count['A']}, T: {nucleodites_count['T']}, C: {nucleodites_count['C']}, G: {nucleodites_count['G']}\"\n",
        "      # Apply colors to the formatted string\n",
        "      colored_counts = self.dna_colors()\n",
        "      formatted_counts = f\"A: {nucleodites_count['A']}, T: {nucleodites_count['T']}, C: {nucleodites_count['C']}, G: {nucleodites_count['G']}\"\n",
        "      return nucleodites_count\n",
        "    else:\n",
        "      nucleodites_count = {'A': 0, 'U': 0, 'C': 0, 'G': 0}\n",
        "      dna = self.dna.upper()\n",
        "      for nuc in dna:\n",
        "          nucleodites_count[nuc] += 1\n",
        "      # Format the output string using the nucleotide counts\n",
        "      #formatted_counts = f\"A: {nucleodites_count['A']}, T: {nucleodites_count['T']}, C: {nucleodites_count['C']}, G: {nucleodites_count['G']}\"\n",
        "      # Apply colors to the formatted string\n",
        "      colored_counts = self.dna_colors()\n",
        "      formatted_counts = f\"A: {nucleodites_count['A']}, U: {nucleodites_count['U']}, C: {nucleodites_count['C']}, G: {nucleodites_count['G']}\"\n",
        "      return nucleodites_count\n",
        "\n",
        " #Return the results\n",
        "\n",
        "  def transcription(self):\n",
        "      \"\"\"Return a list of transcription of the dna\"\"\"\n",
        "      if self.seq_type_ == \"DNA\":\n",
        "           #return self.dna.upper().replace('T','U')\n",
        "           return f\"RNA: \\n 5' {self.dna.upper().replace('T','U')} 3' \\n\"+f\"    {''.join(['|' for nuc in range(len(self.dna))])}\"\n",
        "      else:\n",
        "           #return self.dna.upper().replace('U','T')\n",
        "           #return f\"RNA: \\n 5' {self.dna.upper().replace('U','T')} 3' \\n\"+f\"    {''.join(['|' for nuc in range(len(self.dna))])}\"\n",
        "           return \"Not a DNA sequence, Please add a DNA sequence to transcription.\"\n",
        "\n",
        "  def reverse_complement(self):\n",
        "      \"\"\"Return a list of nucleodites to complete the dna\"\"\"\n",
        "      dna = self.dna.upper()\n",
        "      if self.seq_type_ == \"DNA\":\n",
        "            mapping = str.maketrans('ATCG', 'TAGC')\n",
        "      else:\n",
        "            mapping = str.maketrans('AUCG', 'UAGC')\n",
        "       #dna.translate(mapping)[::-1]\n",
        "      return f\"DNA: \\n 3' {dna.translate(mapping)[::-1]} 5' \\n\"+f\"    {''.join(['|' for nuc in range(len(self.dna))])}\"\n",
        "\n",
        "  def dna_complement(self):\n",
        "      \"\"\"Return a list of nucleodites reverse of a list of nucleodites\"\"\"\n",
        "      #return ''.join([reverse_nucleodites[nuc.upper()] for nuc in self.dna])\n",
        "      return (f\"DNA Complement: \\n 5' {''.join([reverse_nucleodites[nuc.upper()] for nuc in self.dna])} 3' \\n\"+\n",
        "      f\"    {''.join(['|' for nuc in range(len(self.dna))])}\\n\"+\n",
        "      f\"    {self.dna}   \")\n",
        "\n",
        "  def gc_content(self):\n",
        "    \"\"\"GC Content in a DNA/RNA sequence\"\"\"\n",
        "    if len(self.dna) == 0:\n",
        "      return 0\n",
        "    return round((self.dna.count('C') + self.dna.count('G')) / len(self.dna) * 100,6)\n",
        "\n",
        "  def gc_content_subsets(self,Fastadict,k=20):\n",
        "    \"\"\"GC Content in a DNA/RNA sub-sequence length k. k=20 by default\"\"\"\n",
        "    dna = self.dna.upper()\n",
        "    props=[]\n",
        "    for i in range(0,len(dna)-k+1,k):\n",
        "      subsets=dna[i:i+k]\n",
        "      self.dna=subsets\n",
        "      props.append(self.gc_content())\n",
        "    self.dna=dna\n",
        "    DictGC={key:gc_content(value) for (key,value) in Fastadict.items()}\n",
        "    maxGC = (max(DictGC, key=DictGC.get))\n",
        "    return (f'{maxGC}\\n{DictGC[maxGC]}')\n",
        "\n",
        "  # List the aminoacids\n",
        "  def translate_dna(self,position_initial=0):\n",
        "      \"\"\"Translates a DNA sequence into an aminoacid sequence\"\"\"\n",
        "      #Check if dna is a string\n",
        "      if isinstance(self.dna, str):\n",
        "          return [DNA_Codons[self.dna[pos:pos + 3]] for pos in range(position_initial, len(self.dna)-2, 3)]\n",
        "      #If dna is not a string (e.g., a list), iterate over its elements and translate them\n",
        "      elif isinstance(self.dna, list):\n",
        "          translated_dna = []\n",
        "          for item in self.dna:\n",
        "              #Check if each element is a string\n",
        "              if isinstance(item, str):\n",
        "                  #Use the original translate_dna function on strings\n",
        "                  translated_dna.extend(self.translate_dna(item, position_initial))\n",
        "              else:\n",
        "                  #Handle cases for non-string elements if necessary\n",
        "                  pass\n",
        "          return translated_dna\n",
        "      else:\n",
        "          return []\n",
        "\n",
        "  #Frequency of a codon and returns the amino acid\n",
        "  def codon_freq(self, aminoacid):\n",
        "        \"\"\"Provides the frequency of each codon encoding a given aminoacid in a DNA sequence\"\"\"\n",
        "        dna = self.dna.upper()\n",
        "        lista=[]\n",
        "        if self.seq_type_ == \"DNA\":\n",
        "          for i in range(0,len(dna)-2,3):\n",
        "            if DNA_Codons[dna[i:i+3]]==aminoacid:\n",
        "              lista.append(dna[i:i+3])\n",
        "\n",
        "        else:\n",
        "          for i in range(0,len(dna)-2,3):\n",
        "            if RNA_Codons[dna[i:i+3]]==aminoacid:\n",
        "              lista.append(dna[i:i+3])\n",
        "\n",
        "        freqdic=dict(Counter(lista))\n",
        "        total_freq=sum(freqdic.values())\n",
        "        for dna in freqdic:\n",
        "            freqdic[dna]=round(freqdic[dna]/total_freq,2)\n",
        "        return freqdic\n",
        "\n",
        "\n",
        "  #In molecular biology, a reading frame is a specific choice out of the possible ways to read the sequence of nucleotides in a nucleic acid (DNA or RNA)\n",
        "  # molecule as a sequence of triplets. Where these triplets equate to amino acids or stop signals during translation, they are called codons.\n",
        "  def reading_frames(self):\n",
        "    \"\"\"Generate reading frames of a dna sequence, including reverse complement\"\"\"\n",
        "    frames=[]\n",
        "    frames.append(self.translate_dna(position_initial=0))\n",
        "    frames.append(self.translate_dna(position_initial=1))\n",
        "    frames.append(self.translate_dna(position_initial=2))\n",
        "    # Store the original DNA sequence\n",
        "    original_dna = self.dna\n",
        "    # Temporarily use the reverse complement for the remaining frames\n",
        "    self.dna = self.dna.upper().translate(str.maketrans('ATCG', 'TAGC'))[::-1]\n",
        "    frames.append(self.translate_dna(position_initial=0))\n",
        "    frames.append(self.translate_dna(position_initial=1))\n",
        "    frames.append(self.translate_dna(position_initial=2))\n",
        "    # Restore the original DNA sequence\n",
        "    self.dna = original_dna\n",
        "    return frames\n",
        "\n",
        "\n",
        "  def proteins(self):\n",
        "    \"\"\"Compute all possible proteins and return a list of possible proteins\"\"\"\n",
        "    proteinas=[]\n",
        "    proteina_atual=[]\n",
        "    for amino in self.dna:\n",
        "      if amino ==\"_\":\n",
        "        if proteina_atual:\n",
        "          for p in proteina_atual:\n",
        "            proteinas.append(''.join(p))\n",
        "            proteina_atual=[]\n",
        "      else:\n",
        "        if amino == \"M\":\n",
        "          proteina_atual.append(\"\")\n",
        "        for i in range(len(proteina_atual)):\n",
        "          proteina_atual[i]+=amino\n",
        "    return proteinas\n",
        "\n",
        "  def frames_proteins(self,inicio=0,final=0,order='False'):\n",
        "    \"\"\"Compute all possible proteins for all reading frames\"\"\"\n",
        "    \"\"\"Protine Search DB: https://www.ncbi.nlm.nih.gov/nuccore/NM_001185097.2\"\"\"\n",
        "    \"\"\"API can be used to pull protein info\"\"\"\n",
        "    if final> inicio:\n",
        "      frames=reading_frames(self.dna[inicio:final])\n",
        "    else:\n",
        "      frames = reading_frames(self.dna)\n",
        "\n",
        "    find_prot=[]\n",
        "    for frame in frames:\n",
        "      proteinas=proteins(frame)\n",
        "      for prot in proteinas:\n",
        "        find_prot.append(prot)\n",
        "    if order=='True':\n",
        "      find_prot.sort(key=len)\n",
        "    return find_prot\n"
      ],
      "metadata": {
        "id": "KLYJ7_sO_k9X"
      },
      "execution_count": 3,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "Fastadict=read_FASTA('GC_TEST.txt')\n",
        "first_sequence_key = list(Fastadict.keys())[0]\n",
        "first_sequence_value = Fastadict[first_sequence_key]\n",
        "dna_random= first_sequence_value"
      ],
      "metadata": {
        "id": "X6wbTOM1ebGV"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "dna_obj = bioinformatics_funcions(dna=dna_random)\n",
        "print('\\n Check if my Data is a DNA Sequence \\n')\n",
        "print(dna_obj.validate_seg_dna())\n",
        "print('\\n My List of nucleodites \\n')\n",
        "print(dna_obj.values()[0])\n",
        "#print(f\"DNA: \\n 5' {dna_obj.values()[0]} 3' \\n\"+f\"    {''.join(['|' for nuc in range(len(dna_obj.values()[0]))])}\")\n",
        "print('\\n Check the Sequence Type \\n')\n",
        "print(dna_obj.seq_type(),\n",
        "dna_obj.seq_info())\n",
        "print('\\n Check the Number of Nucleodites I have in my DNA list \\n')\n",
        "print(dna_obj.count_nucleodites())\n",
        "#print(f\"DNA + Reverse: \\n 5' {dna_colors(dna_random)} 3'\")\n",
        "#print(f\"    {''.join(['|' for nuc in range(len(dna_random))])}\")\n",
        "print('\\n Transcription of DNA to RNA \\n')\n",
        "print(dna_obj.transcription())\n",
        "#print(f\"RNA: \\n 5' {dna_obj.transcription()} 3' \\n\"+f\"    {''.join(['|' for nuc in range(len(dna_obj.transcription()))])}\")\n",
        "print('\\n Implement my DNA  Reverse Complement \\n')\n",
        "#print(f\"DNA: \\n 3' {dna_obj.reverse_complement()} 5' \\n\"+f\"    {''.join(['|' for nuc in range(len(dna_obj.reverse_complement()))])}\")\n",
        "print(dna_obj.reverse_complement())\n",
        "print('\\n Implement my DNA complement \\n')\n",
        "#print(f\"DNA Complement: \\n 5' {dna_obj.dna_complement()} 3' \\n\"+f\"    {''.join(['|' for nuc in range(len(dna_obj.dna_complement()))])}\")\n",
        "print(dna_obj.dna_complement())\n",
        "print('\\n Calculate my GC Content total and subsets \\n')\n",
        "print(dna_obj.gc_content())\n",
        "print(dna_obj.gc_content_subsets(Fastadict))\n",
        "print(f'\\n Amino Acid Sequence \\n')\n",
        "print(dna_obj.translate_dna())\n",
        "print(f'\\n Frequency of a codon and returns the amino acid \\n')\n",
        "print(dna_obj.codon_freq('P'))\n",
        "print(f'\\n Reading Frames \\n')\n",
        "for reading in dna_obj.reading_frames():\n",
        "  print(reading)\n",
        "print(f'\\n List of Proteins \\n')\n",
        "for protein in dna_obj.frames_proteins():\n",
        "  print(protein)\n",
        "\n",
        "# writing a file, read a file\n",
        "#writetextfile('test.txt',dna_obj,'a') - 'a  indicate append the elements\n",
        "#readTextFile('test.txt')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "uCzcQSCIej14",
        "outputId": "8e7749de-ea7f-406a-c47c-5f1b221c827f"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            " Check if my Data is a DNA Sequence \n",
            "\n",
            "True\n",
            "\n",
            " My List of nucleodites \n",
            "\n",
            "DNA: \n",
            " 5' AGCCCTCCAGGACAGGCTGCATCAGAAGAGGCCATCAAGCAGGTCTGTTCCAAGGGCCTTTGCGTCAGATCACTGTCCTTCTGCCATGGCCCTGTGGATGCGCCTCCTGCCCCTGCTGGCGCTGCTGGCCCTCTGGGGACCTGACCCAGCCGCAGCCTTTGTGAACCAACACCTGTGCGGCTCACACCTGGTGGAAGCTCTCTACCTAGTGTGCGGGGAACGAGGCTTCTTCTACACACCCAAGACCCGCCGGGAGGCAGAGGACCTGCAGGTGGGGCAGGTGGAGCTGGGCGGGGGCCCTGGTGCAGGCAGCCTGCAGCCCTTGGCCCTGGAGGGGTCCCTGCAGAAGCGTGGCATTGTGGAACAATGCTGTACCAGCATCTGCTCCCTCTACCAGCTGGAGAACTACTGCAACTAGACGCAGCCCGCAGGCAGCCCCACACCCGCCGCCTCCTGCACCGAGAGAGATGGAATAAAGCCCTTGAACCAGC 3' \n",
            "    |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
            "\n",
            " Check the Sequence Type \n",
            "\n",
            "DNA [Label]: No Label\n",
            "[Sequence]: AGCCCTCCAGGACAGGCTGCATCAGAAGAGGCCATCAAGCAGGTCTGTTCCAAGGGCCTTTGCGTCAGATCACTGTCCTTCTGCCATGGCCCTGTGGATGCGCCTCCTGCCCCTGCTGGCGCTGCTGGCCCTCTGGGGACCTGACCCAGCCGCAGCCTTTGTGAACCAACACCTGTGCGGCTCACACCTGGTGGAAGCTCTCTACCTAGTGTGCGGGGAACGAGGCTTCTTCTACACACCCAAGACCCGCCGGGAGGCAGAGGACCTGCAGGTGGGGCAGGTGGAGCTGGGCGGGGGCCCTGGTGCAGGCAGCCTGCAGCCCTTGGCCCTGGAGGGGTCCCTGCAGAAGCGTGGCATTGTGGAACAATGCTGTACCAGCATCTGCTCCCTCTACCAGCTGGAGAACTACTGCAACTAGACGCAGCCCGCAGGCAGCCCCACACCCGCCGCCTCCTGCACCGAGAGAGATGGAATAAAGCCCTTGAACCAGC\n",
            "[Type]: DNA\n",
            "[Length]: 491\n",
            "\n",
            " Check the Number of Nucleodites I have in my DNA list \n",
            "\n",
            "{'A': 94, 'T': 85, 'C': 163, 'G': 149}\n",
            "\n",
            " Transcription of DNA to RNA \n",
            "\n",
            "RNA: \n",
            " 5' AGCCCUCCAGGACAGGCUGCAUCAGAAGAGGCCAUCAAGCAGGUCUGUUCCAAGGGCCUUUGCGUCAGAUCACUGUCCUUCUGCCAUGGCCCUGUGGAUGCGCCUCCUGCCCCUGCUGGCGCUGCUGGCCCUCUGGGGACCUGACCCAGCCGCAGCCUUUGUGAACCAACACCUGUGCGGCUCACACCUGGUGGAAGCUCUCUACCUAGUGUGCGGGGAACGAGGCUUCUUCUACACACCCAAGACCCGCCGGGAGGCAGAGGACCUGCAGGUGGGGCAGGUGGAGCUGGGCGGGGGCCCUGGUGCAGGCAGCCUGCAGCCCUUGGCCCUGGAGGGGUCCCUGCAGAAGCGUGGCAUUGUGGAACAAUGCUGUACCAGCAUCUGCUCCCUCUACCAGCUGGAGAACUACUGCAACUAGACGCAGCCCGCAGGCAGCCCCACACCCGCCGCCUCCUGCACCGAGAGAGAUGGAAUAAAGCCCUUGAACCAGC 3' \n",
            "    |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
            "\n",
            " Implement my DNA  Reverse Complement \n",
            "\n",
            "DNA: \n",
            " 3' GCTGGTTCAAGGGCTTTATTCCATCTCTCTCGGTGCAGGAGGCGGCGGGTGTGGGGCTGCCTGCGGGCTGCGTCTAGTTGCAGTAGTTCTCCAGCTGGTAGAGGGAGCAGATGCTGGTACAGCATTGTTCCACAATGCCACGCTTCTGCAGGGACCCCTCCAGGGCCAAGGGCTGCAGGCTGCCTGCACCAGGGCCCCCGCCCAGCTCCACCTGCCCCACCTGCAGGTCCTCTGCCTCCCGGCGGGTCTTGGGTGTGTAGAAGAAGCCTCGTTCCCCGCACACTAGGTAGAGAGCTTCCACCAGGTGTGAGCCGCACAGGTGTTGGTTCACAAAGGCTGCGGCTGGGTCAGGTCCCCAGAGGGCCAGCAGCGCCAGCAGGGGCAGGAGGCGCATCCACAGGGCCATGGCAGAAGGACAGTGATCTGACGCAAAGGCCCTTGGAACAGACCTGCTTGATGGCCTCTTCTGATGCAGCCTGTCCTGGAGGGCT 5' \n",
            "    |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
            "\n",
            " Implement my DNA complement \n",
            "\n",
            "DNA Complement: \n",
            " 5' TCGGGAGGTCCTGTCCGACGTAGTCTTCTCCGGTAGTTCGTCCAGACAAGGTTCCCGGAAACGCAGTCTAGTGACAGGAAGACGGTACCGGGACACCTACGCGGAGGACGGGGACGACCGCGACGACCGGGAGACCCCTGGACTGGGTCGGCGTCGGAAACACTTGGTTGTGGACACGCCGAGTGTGGACCACCTTCGAGAGATGGATCACACGCCCCTTGCTCCGAAGAAGATGTGTGGGTTCTGGGCGGCCCTCCGTCTCCTGGACGTCCACCCCGTCCACCTCGACCCGCCCCCGGGACCACGTCCGTCGGACGTCGGGAACCGGGACCTCCCCAGGGACGTCTTCGCACCGTAACACCTTGTTACGACATGGTCGTAGACGAGGGAGATGGTCGACCTCTTGATGACGTTGATCTGCGTCGGGCGTCCGTCGGGGTGTGGGCGGCGGAGGACGTGGCTCTCTCTACCTTATTTCGGGAACTTGGTCG 3' \n",
            "    |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
            "    AGCCCTCCAGGACAGGCTGCATCAGAAGAGGCCATCAAGCAGGTCTGTTCCAAGGGCCTTTGCGTCAGATCACTGTCCTTCTGCCATGGCCCTGTGGATGCGCCTCCTGCCCCTGCTGGCGCTGCTGGCCCTCTGGGGACCTGACCCAGCCGCAGCCTTTGTGAACCAACACCTGTGCGGCTCACACCTGGTGGAAGCTCTCTACCTAGTGTGCGGGGAACGAGGCTTCTTCTACACACCCAAGACCCGCCGGGAGGCAGAGGACCTGCAGGTGGGGCAGGTGGAGCTGGGCGGGGGCCCTGGTGCAGGCAGCCTGCAGCCCTTGGCCCTGGAGGGGTCCCTGCAGAAGCGTGGCATTGTGGAACAATGCTGTACCAGCATCTGCTCCCTCTACCAGCTGGAGAACTACTGCAACTAGACGCAGCCCGCAGGCAGCCCCACACCCGCCGCCTCCTGCACCGAGAGAGATGGAATAAAGCCCTTGAACCAGC   \n",
            "\n",
            " Calculate my GC Content total and subsets \n",
            "\n",
            "63.543788\n",
            ">NM_001185097.2 Homo sapiens insulin (INS), transcript variant 2, mRNA\n",
            "63.543788\n",
            "\n",
            " Amino Acid Sequence \n",
            "\n",
            "['S', 'P', 'P', 'G', 'Q', 'A', 'A', 'S', 'E', 'E', 'A', 'I', 'K', 'Q', 'V', 'C', 'S', 'K', 'G', 'L', 'C', 'V', 'R', 'S', 'L', 'S', 'F', 'C', 'H', 'G', 'P', 'V', 'D', 'A', 'P', 'P', 'A', 'P', 'A', 'G', 'A', 'A', 'G', 'P', 'L', 'G', 'T', '_', 'P', 'S', 'R', 'S', 'L', 'C', 'E', 'P', 'T', 'P', 'V', 'R', 'L', 'T', 'P', 'G', 'G', 'S', 'S', 'L', 'P', 'S', 'V', 'R', 'G', 'T', 'R', 'L', 'L', 'L', 'H', 'T', 'Q', 'D', 'P', 'P', 'G', 'G', 'R', 'G', 'P', 'A', 'G', 'G', 'A', 'G', 'G', 'A', 'G', 'R', 'G', 'P', 'W', 'C', 'R', 'Q', 'P', 'A', 'A', 'L', 'G', 'P', 'G', 'G', 'V', 'P', 'A', 'E', 'A', 'W', 'H', 'C', 'G', 'T', 'M', 'L', 'Y', 'Q', 'H', 'L', 'L', 'P', 'L', 'P', 'A', 'G', 'E', 'L', 'L', 'Q', 'L', 'D', 'A', 'A', 'R', 'R', 'Q', 'P', 'H', 'T', 'R', 'R', 'L', 'L', 'H', 'R', 'E', 'R', 'W', 'N', 'K', 'A', 'L', 'E', 'P']\n",
            "\n",
            " Frequency of a codon and returns the amino acid \n",
            "\n",
            "{'CCT': 0.61, 'CCA': 0.17, 'CCC': 0.13, 'CCG': 0.09}\n",
            "\n",
            " Reading Frames \n",
            "\n",
            "['S', 'P', 'P', 'G', 'Q', 'A', 'A', 'S', 'E', 'E', 'A', 'I', 'K', 'Q', 'V', 'C', 'S', 'K', 'G', 'L', 'C', 'V', 'R', 'S', 'L', 'S', 'F', 'C', 'H', 'G', 'P', 'V', 'D', 'A', 'P', 'P', 'A', 'P', 'A', 'G', 'A', 'A', 'G', 'P', 'L', 'G', 'T', '_', 'P', 'S', 'R', 'S', 'L', 'C', 'E', 'P', 'T', 'P', 'V', 'R', 'L', 'T', 'P', 'G', 'G', 'S', 'S', 'L', 'P', 'S', 'V', 'R', 'G', 'T', 'R', 'L', 'L', 'L', 'H', 'T', 'Q', 'D', 'P', 'P', 'G', 'G', 'R', 'G', 'P', 'A', 'G', 'G', 'A', 'G', 'G', 'A', 'G', 'R', 'G', 'P', 'W', 'C', 'R', 'Q', 'P', 'A', 'A', 'L', 'G', 'P', 'G', 'G', 'V', 'P', 'A', 'E', 'A', 'W', 'H', 'C', 'G', 'T', 'M', 'L', 'Y', 'Q', 'H', 'L', 'L', 'P', 'L', 'P', 'A', 'G', 'E', 'L', 'L', 'Q', 'L', 'D', 'A', 'A', 'R', 'R', 'Q', 'P', 'H', 'T', 'R', 'R', 'L', 'L', 'H', 'R', 'E', 'R', 'W', 'N', 'K', 'A', 'L', 'E', 'P']\n",
            "['A', 'L', 'Q', 'D', 'R', 'L', 'H', 'Q', 'K', 'R', 'P', 'S', 'S', 'R', 'S', 'V', 'P', 'R', 'A', 'F', 'A', 'S', 'D', 'H', 'C', 'P', 'S', 'A', 'M', 'A', 'L', 'W', 'M', 'R', 'L', 'L', 'P', 'L', 'L', 'A', 'L', 'L', 'A', 'L', 'W', 'G', 'P', 'D', 'P', 'A', 'A', 'A', 'F', 'V', 'N', 'Q', 'H', 'L', 'C', 'G', 'S', 'H', 'L', 'V', 'E', 'A', 'L', 'Y', 'L', 'V', 'C', 'G', 'E', 'R', 'G', 'F', 'F', 'Y', 'T', 'P', 'K', 'T', 'R', 'R', 'E', 'A', 'E', 'D', 'L', 'Q', 'V', 'G', 'Q', 'V', 'E', 'L', 'G', 'G', 'G', 'P', 'G', 'A', 'G', 'S', 'L', 'Q', 'P', 'L', 'A', 'L', 'E', 'G', 'S', 'L', 'Q', 'K', 'R', 'G', 'I', 'V', 'E', 'Q', 'C', 'C', 'T', 'S', 'I', 'C', 'S', 'L', 'Y', 'Q', 'L', 'E', 'N', 'Y', 'C', 'N', '_', 'T', 'Q', 'P', 'A', 'G', 'S', 'P', 'T', 'P', 'A', 'A', 'S', 'C', 'T', 'E', 'R', 'D', 'G', 'I', 'K', 'P', 'L', 'N', 'Q']\n",
            "['P', 'S', 'R', 'T', 'G', 'C', 'I', 'R', 'R', 'G', 'H', 'Q', 'A', 'G', 'L', 'F', 'Q', 'G', 'P', 'L', 'R', 'Q', 'I', 'T', 'V', 'L', 'L', 'P', 'W', 'P', 'C', 'G', 'C', 'A', 'S', 'C', 'P', 'C', 'W', 'R', 'C', 'W', 'P', 'S', 'G', 'D', 'L', 'T', 'Q', 'P', 'Q', 'P', 'L', '_', 'T', 'N', 'T', 'C', 'A', 'A', 'H', 'T', 'W', 'W', 'K', 'L', 'S', 'T', '_', 'C', 'A', 'G', 'N', 'E', 'A', 'S', 'S', 'T', 'H', 'P', 'R', 'P', 'A', 'G', 'R', 'Q', 'R', 'T', 'C', 'R', 'W', 'G', 'R', 'W', 'S', 'W', 'A', 'G', 'A', 'L', 'V', 'Q', 'A', 'A', 'C', 'S', 'P', 'W', 'P', 'W', 'R', 'G', 'P', 'C', 'R', 'S', 'V', 'A', 'L', 'W', 'N', 'N', 'A', 'V', 'P', 'A', 'S', 'A', 'P', 'S', 'T', 'S', 'W', 'R', 'T', 'T', 'A', 'T', 'R', 'R', 'S', 'P', 'Q', 'A', 'A', 'P', 'H', 'P', 'P', 'P', 'P', 'A', 'P', 'R', 'E', 'M', 'E', '_', 'S', 'P', '_', 'T', 'S']\n",
            "['A', 'G', 'S', 'R', 'A', 'L', 'F', 'H', 'L', 'S', 'R', 'C', 'R', 'R', 'R', 'R', 'V', 'W', 'G', 'C', 'L', 'R', 'A', 'A', 'S', 'S', 'C', 'S', 'S', 'S', 'P', 'A', 'G', 'R', 'G', 'S', 'R', 'C', 'W', 'Y', 'S', 'I', 'V', 'P', 'Q', 'C', 'H', 'A', 'S', 'A', 'G', 'T', 'P', 'P', 'G', 'P', 'R', 'A', 'A', 'G', 'C', 'L', 'H', 'Q', 'G', 'P', 'R', 'P', 'A', 'P', 'P', 'A', 'P', 'P', 'A', 'G', 'P', 'L', 'P', 'P', 'G', 'G', 'S', 'W', 'V', 'C', 'R', 'R', 'S', 'L', 'V', 'P', 'R', 'T', 'L', 'G', 'R', 'E', 'L', 'P', 'P', 'G', 'V', 'S', 'R', 'T', 'G', 'V', 'G', 'S', 'Q', 'R', 'L', 'R', 'L', 'G', 'Q', 'V', 'P', 'R', 'G', 'P', 'A', 'A', 'P', 'A', 'G', 'A', 'G', 'G', 'A', 'S', 'T', 'G', 'P', 'W', 'Q', 'K', 'D', 'S', 'D', 'L', 'T', 'Q', 'R', 'P', 'L', 'E', 'Q', 'T', 'C', 'L', 'M', 'A', 'S', 'S', 'D', 'A', 'A', 'C', 'P', 'G', 'G']\n",
            "['L', 'V', 'Q', 'G', 'L', 'Y', 'S', 'I', 'S', 'L', 'G', 'A', 'G', 'G', 'G', 'G', 'C', 'G', 'A', 'A', 'C', 'G', 'L', 'R', 'L', 'V', 'A', 'V', 'V', 'L', 'Q', 'L', 'V', 'E', 'G', 'A', 'D', 'A', 'G', 'T', 'A', 'L', 'F', 'H', 'N', 'A', 'T', 'L', 'L', 'Q', 'G', 'P', 'L', 'Q', 'G', 'Q', 'G', 'L', 'Q', 'A', 'A', 'C', 'T', 'R', 'A', 'P', 'A', 'Q', 'L', 'H', 'L', 'P', 'H', 'L', 'Q', 'V', 'L', 'C', 'L', 'P', 'A', 'G', 'L', 'G', 'C', 'V', 'E', 'E', 'A', 'S', 'F', 'P', 'A', 'H', '_', 'V', 'E', 'S', 'F', 'H', 'Q', 'V', '_', 'A', 'A', 'Q', 'V', 'L', 'V', 'H', 'K', 'G', 'C', 'G', 'W', 'V', 'R', 'S', 'P', 'E', 'G', 'Q', 'Q', 'R', 'Q', 'Q', 'G', 'Q', 'E', 'A', 'H', 'P', 'Q', 'G', 'H', 'G', 'R', 'R', 'T', 'V', 'I', '_', 'R', 'K', 'G', 'P', 'W', 'N', 'R', 'P', 'A', '_', 'W', 'P', 'L', 'L', 'M', 'Q', 'P', 'V', 'L', 'E', 'G']\n",
            "['W', 'F', 'K', 'G', 'F', 'I', 'P', 'S', 'L', 'S', 'V', 'Q', 'E', 'A', 'A', 'G', 'V', 'G', 'L', 'P', 'A', 'G', 'C', 'V', '_', 'L', 'Q', '_', 'F', 'S', 'S', 'W', '_', 'R', 'E', 'Q', 'M', 'L', 'V', 'Q', 'H', 'C', 'S', 'T', 'M', 'P', 'R', 'F', 'C', 'R', 'D', 'P', 'S', 'R', 'A', 'K', 'G', 'C', 'R', 'L', 'P', 'A', 'P', 'G', 'P', 'P', 'P', 'S', 'S', 'T', 'C', 'P', 'T', 'C', 'R', 'S', 'S', 'A', 'S', 'R', 'R', 'V', 'L', 'G', 'V', '_', 'K', 'K', 'P', 'R', 'S', 'P', 'H', 'T', 'R', '_', 'R', 'A', 'S', 'T', 'R', 'C', 'E', 'P', 'H', 'R', 'C', 'W', 'F', 'T', 'K', 'A', 'A', 'A', 'G', 'S', 'G', 'P', 'Q', 'R', 'A', 'S', 'S', 'A', 'S', 'R', 'G', 'R', 'R', 'R', 'I', 'H', 'R', 'A', 'M', 'A', 'E', 'G', 'Q', '_', 'S', 'D', 'A', 'K', 'A', 'L', 'G', 'T', 'D', 'L', 'L', 'D', 'G', 'L', 'F', '_', 'C', 'S', 'L', 'S', 'W', 'R', 'A']\n",
            "\n",
            " List of Proteins \n",
            "\n",
            "MALWMRLLPLLALLALWGPDPAAAFVNQHLCGSHLVEALYLVCGERGFFYTPKTRREAEDLQVGQVELGGGPGAGSLQPLALEGSLQKRGIVEQCCTSICSLYQLENYCN\n",
            "MRLLPLLALLALWGPDPAAAFVNQHLCGSHLVEALYLVCGERGFFYTPKTRREAEDLQVGQVELGGGPGAGSLQPLALEGSLQKRGIVEQCCTSICSLYQLENYCN\n",
            "ME\n",
            "MLVQHCSTMPRFCRDPSRAKGCRLPAPGPPPSSTCPTCRSSASRRVLGV\n",
            "MPRFCRDPSRAKGCRLPAPGPPPSSTCPTCRSSASRRVLGV\n",
            "MAEGQ\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def hamming_dist(dna1,dna2):\n",
        "    zip_dna=zip(dna1,dna2)\n",
        "    h_dist=[(n1,n2) for n1,n2 in zip_dna if n1!=n2]\n",
        "    return f\"The Hamming Distance is {len(h_dist)} unitis\"\n",
        "hamming_dist('TCGAAA','TAGCAA')"
      ],
      "metadata": {
        "id": "05borYfRpEDU",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 35
        },
        "outputId": "7369983d-d2d4-4bb1-855a-9ca6eac85663"
      },
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "'The Hamming Distance is 2 unitis'"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 10
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Genomic  Projetc"
      ],
      "metadata": {
        "id": "gb9hCJMOc4TP"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "class genome_python:\n",
        "  def __init__(self):\n",
        "    print('Genome class')\n",
        "\n",
        "  def count_kmers(self, dna, kmer):\n",
        "    '''We can define a k-mer as a substring of length k of a DNA, RNA or protein sequence.\n",
        "       A sequence of length n can be constructed by overlapping n - k + 1 k-mers'''\n",
        "    kmer_count = 0\n",
        "    for i in range(len(dna) -1):\n",
        "        if dna[i:i+len(kmer)] == kmer:\n",
        "            kmer_count += 1\n",
        "    return kmer_count\n",
        "\n",
        "  def find_frequent_kmers(self, dna, k_len):\n",
        "        \"\"\"Find the most frequent k-mers in a sequence.\"\"\"\n",
        "        kmer_freq = {}\n",
        "\n",
        "        for i in range(len(dna) - k_len + 1):\n",
        "            kmer = dna[i:i+k_len]\n",
        "            if kmer in kmer_freq:\n",
        "                kmer_freq[kmer] += 1\n",
        "            else:\n",
        "                kmer_freq[kmer] = 1\n",
        "\n",
        "        highest_frequency = max(kmer_freq.values())\n",
        "\n",
        "        return [\n",
        "            kmer for kmer, frequency in kmer_freq.items()\n",
        "            if frequency == highest_frequency]\n"
      ],
      "metadata": {
        "id": "0pgUHTxyc3St"
      },
      "execution_count": 24,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "dna='ATCGATCGATCGATCGACTGACTAG'\n",
        "kmer='TC'\n",
        "tamanho=3\n",
        "genome=genome_python()\n",
        "print(f'\\n The total of {kmer} repetions \\n')\n",
        "print(genome.count_kmers(dna,kmer))\n",
        "print(f'\\n The most Frequent kmers with lengh {tamanho} \\n')\n",
        "print(genome.find_frequent_kmers(dna,tamanho))"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "3HgOHx1Ec3Vd",
        "outputId": "404d6c5f-2d88-475c-fa9e-ed60bc9c3a6c"
      },
      "execution_count": 29,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Genome class\n",
            "\n",
            " The total of TC repetions \n",
            "\n",
            "4\n",
            "\n",
            " The most Frequent kmers with lengh 3 \n",
            "\n",
            "['ATC', 'TCG', 'CGA']\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "xRic4PQZc3YO"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "Ita9HyVyc3ax"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}